collections/src/main/scala/strawman/collection/immutable/ImmutableArray.scala

```diff
@@ -152,7 +151,140 @@ object ImmutableArray extends StrictOptimizedSeqFactory[ImmutableArray] {
       ScalaRunTime.array_update(elements, i, f(i))
       i = i + 1
     }
-    new ImmutableArray(elements)
+    ImmutableArray.wrapArray(elements)
+  }
+
+  def wrapArray[T](x: AnyRef): ImmutableArray[T] = (x match {
+    case null              => null
+    case x: Array[AnyRef]  => new ofRef[AnyRef](x)
+    case x: Array[Int]     => new ofInt(x)
+    case x: Array[Double]  => new ofDouble(x)
+    case x: Array[Long]    => new ofLong(x)
+    case x: Array[Float]   => new ofFloat(x)
+    case x: Array[Char]    => new ofChar(x)
+    case x: Array[Byte]    => new ofByte(x)
+    case x: Array[Short]   => new ofShort(x)
+    case x: Array[Boolean] => new ofBoolean(x)
+    case x: Array[Unit]    => new ofUnit(x)
+  }).asInstanceOf[ImmutableArray[T]]
+
+  final class ofRef[T <: AnyRef](val array: Array[T]) extends ImmutableArray[T] with Serializable {
```

Comment:
I think we should follow the well established naming convention where class names begin with an upper case letter.

Reply:
They're public and copied directly from the current API. I wouldn't want to do any unnecessary breaking changes here.